components/batch-system/src/batch.rs

```diff
@@ -329,7 +329,8 @@ impl<N: Fsm, C: Fsm, Handler: PollHandler<N, C>> Poller<N, C, Handler> {
                 }
                 fsm_cnt += 1;
             }
-            self.handler.end(&mut batch.normals);
+            self.handler
+                .end(&mut batch.normals, batch.control.as_mut().unwrap());
```

Comment:
This is not a good idea to deal with `control_fsm` at `end`, because handler has released it before.